Fastapi scoring

shubh0508 opened the pull request:
## What changes are proposed in this pull request?

PR for FastAPI based scoring server [discussion](https://github.com/mlflow/mlflow/discussions/5601).

## How is this patch tested?

(Details)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

mlflow-automation committed:
Update MLflow version to 1.22.0 (#5123)

Signed-off-by: Jenkins <jenkins@databricks.com>

Co-authored-by: Jenkins <jenkins@databricks.com>
Signed-off-by: shubh chaurasia <shubhchaurasia.iitd@gmail.com>

shubh0508 commented:
Dummy PR for this [discussion](https://github.com/mlflow/mlflow/discussions/5601).

dbczumar commented:
Hi @shubh0508 , thanks for investigating! Would you be able to further instrument the MLflow model serving procedure to identify the bottleneck? Adding some latency measurements (e.g. by using `time.time()` and comparing timestamps before / after certain steps) between request send, request receipt, request message parsing completion, model prediction start, model prediction end, etc. might help to narrow things down.

shubh0508 commented:
Hi @dbczumar, I found the issue with the FastAPI scoring server. 
With PyfuncModel objects, I need to use `def transformation` instead of `async def transformation`. I think it is so because PyfuncModel objects don't seem to be compatible with the new **asyncio**  module APIs. I will look into it to find the exact cause. But for now, I will be able to use FastAPI with MLFlow for our production deployment.

@dbczumar Don't we have any plans to replace Flask with FastAPI in MLFlow?

shubh0508 committed:
FastAPI model changes

Signed-off-by: shubh chaurasia <shubhchaurasia.iitd@gmail.com>

shubh0508 commented:
@harupy Do we have plans to use the FastAPI scoring server instead of the Flask server?
Should I work on replacing Flask based scoring ?

robsonpeixoto commented:
Great job, @shubh0508 :heart: 
Did you do any benchmark? The performance was better? And latency?

shubh0508 commented:
Yes @robsonpeixoto. I had done the benchmark. The performance was much better in terms of latency and throughput.

So, when I started on the task, a single inference GBM model serving was taking around 8ms-10ms. Then I looked into the inference process and did following optimisations.

Default implementation uses pandas Dataframe as standard request format for MLFlow model, but creating a Dataframe usually takes around 4ms-5ms. So, I had replaced it with Pydantic raw-json request handler. 

Removing dataframe request format brought down the latency to 3.5ms-5ms.

Then, I replaced the Flask app with FastAPI and it brought the latency to 1.5ms-3ms. 

So, I implemented this change across 50+ models running on Zomato production systems and brought down the infra costs by atleast 50%. Some of the models were running with over a million RPM.
